src/java.base/share/classes/javax/crypto/KDFSpi.java

```diff
@@ -0,0 +1,149 @@
+ * questions.
+ */
+
+package javax.crypto;
+
+import jdk.internal.javac.PreviewFeature;
+
+import java.security.InvalidAlgorithmParameterException;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.AlgorithmParameterSpec;
+
+/**
+ * This class defines the <i>Service Provider Interface</i> (<b>SPI</b>) for the
+ * Key Derivation Function ({@link KDF}) class.
+ * <p>
+ * All the abstract methods in this class must be implemented by each
+ * cryptographic service provider who wishes to supply the implementation of a
+ * particular key derivation algorithm.
+ * <p>
+ * In addition, all implementations must provide a public constructor which
+ * accepts a {@code KDFParameters} object. The constructor must call {@code
+ * super(params)} passing the parameters supplied. The constructor must also
+ * throw an {@code InvalidAlgorithmParameterException} if the supplied
+ * parameters are inappropriate.
+ * <p>
+ * Implementations which do not support {@code KDFParameters} must require
+ * {@code null} to be passed, otherwise an {@code InvalidAlgorithmParameterException}
+ * will be thrown. On the other hand, implementations which require
+ * {@code KDFParameters} should throw an {@code InvalidAlgorithmParameterException}
+ * upon receiving a {@code null} value or {@code KDFParameters} which are not
+ * supported by the implementation.
+ * <p>
+ * To aid the caller, implementations may return parameters with additional
+ * default values or supply random values as used by the underlying {@code KDF}
+ * algorithm. See {@link KDFSpi#engineGetParameters()} for more details.
+ *
+ * @see KDF
+ * @see KDFParameters
+ * @see KDF#getParameters()
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public abstract class KDFSpi {
+
+    /**
+     * The sole constructor.
+     * <p>
+     * A {@code KDFParameters} object may be specified for KDF algorithms
+     * that support initialization parameters.
+     *
+     * @param kdfParameters
+     *     the initialization parameters for the {@code KDF} algorithm (may be
+     *     {@code null})
+     *
+     * @see KDF#getParameters()
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDFSpi}
+     */
+    protected KDFSpi(KDFParameters kdfParameters)
+        throws InvalidAlgorithmParameterException {}
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
```

Comment:
According to the class spec, when you say parameters are _required_ the caller must provide a non-null value. The sentence above contradicts with this.

Reply:
Your previous sentence contains a very important information that even if no parameters were provided this method might still return one. Now it's only "certain fields of the parameters". This is not as strong as the previous one.